libsolidity/analysis/TypeChecker.cpp

```diff
@@ -596,19 +596,32 @@ bool TypeChecker::visit(VariableDeclaration const& _variable)
 	if (isStructMemberDeclaration)
 		return false;
 
-	if (auto referenceType = dynamic_cast<ReferenceType const*>(varType))
+	auto compositeType = dynamic_cast<CompositeType const*>(varType);
+	if (compositeType == nullptr)
+		return false;
+
+	bool callDataCheckRequired = true;
+	if (_variable.isLibraryFunctionParameter())
+		callDataCheckRequired = false;
+	else if (_variable.isEventParameter() && _variable.annotation().type->containsNestedMapping())
 	{
-		auto result = referenceType->validForLocation(referenceType->location());
-		if (result && (_variable.isConstructorParameter() || _variable.isPublicCallableParameter()))
-			result = referenceType->validForLocation(DataLocation::CallData);
-		if (!result)
-		{
-			solAssert(!result.message().empty(), "Expected detailed error message");
-			m_errorReporter.typeError(1534_error, _variable.location(), result.message());
-			return false;
-		}
+		solAssert(
+			m_errorReporter.errorCount() > 0,
+			"Expected \"Type containing a (nested) mapping is not allowed as event parameter type.\""
+		);
+		callDataCheckRequired = false;
 	}
```

Comment:
`indexed` event parameters definitely should be skipped. Not sure about others. They get ABI encoded but they're not stored in calldata and the space does not have a hard limit. But calldata/memory do not have an official limit either (at least not yet: #9810) and we still enforce the size limit there. One difference is that the space for logs isn't addressable from within Solidity so it's not implicitly limited by pointer size.

Relevant docs: https://solidity.readthedocs.io/en/v0.7.1/contracts.html#events
> All parameters without the indexed attribute are ABI-encoded into the data part of the log.

Reply:
Done. (BTW, I did not squash the changes yet, to simplify the review.)

Reply:
You're now only skipping indexed parameters containing mappings. Actually, you're skipping all event parameters, but I think that's unintentional (https://github.com/ethereum/solidity/pull/9746#discussion_r496117101).

Anyway, what do you think about checking the non-indexed ones? Should they follow `calldata` rules too? I'm not sure myself, that's why I'm asking... They're technically never stored in calldata but not validating them at all seems wrong too.

Reply:
Should we resolve this conversation and continue in https://github.com/ethereum/solidity/pull/9746#discussion_r496117101? Or should I just close this PR and make a new one, to get rid of old threads and simplify the discussion?

Reply:
I'm not sure about a clean PR. It would probably get you more reviews since this one is long and intimidating at this point. On the other hand, almost all comments are resolved at this point and the remaining ones all suggestions, my comments about adding more tests and the one or two points we're still discussing :)

I guess I don't mind either way. I think we're pretty close to being done here but you do need more input from others here.